tutorials/flappy-bird/step_8/flappy-bird.cpp

```diff
@@ -0,0 +1,660 @@
+
+            // Size X is full canvas but the height is defined in constants
+            // We also make it a bit longer by adding the thickness of the outline to hide the outline at sides
+            math::vec2f size{canvas_width + constants.grass_outline_color.thickness * 2.0f, constants.grass_thickness};
+
+            auto grass = geometry::blueprint_rectangle(registry, size, constants.grass_color, pos,
+                                                       constants.grass_outline_color);
+            registry.assign<graphics::layer<3>>(grass);
+            registry.assign<entt::tag<"game_scene"_hs>>(grass);
+        }
+
+        // Create Ground
+        {
+            // Ground expands to whole canvas width so position is middle of it,
+            // But position Y is at bottom of the screen so it's full canvas_height minus half of the ground thickness
+            transform::position_2d pos{canvas_width * 0.5f, canvas_height - constants.ground_thickness * 0.5f};
+
+            // Size X is full canvas but the height is defined in constants
+            math::vec2f size{canvas_width, constants.ground_thickness};
+
+            auto ground = geometry::blueprint_rectangle(registry, size, constants.ground_color, pos);
+            registry.assign<graphics::layer<3>>(ground);
+            registry.assign<entt::tag<"game_scene"_hs>>(ground);
+        }
+    }
+
+    // Factory for creating the player
+    entt::entity create_player(entt::registry &registry) {
+        // Retrieve constants
+        const auto[_, canvas_height] = registry.ctx<graphics::canvas_2d>().canvas.size;
+        const auto constants = registry.ctx<flappy_bird_constants>();
+
+        auto entity = graphics::blueprint_sprite(registry,
+                                                 graphics::sprite{constants.player_image_name.c_str()},
+                                                 transform::position_2d{constants.player_pos_x, canvas_height * 0.5f});
+        registry.assign<antara::gaming::graphics::layer<5>>(entity);
+        registry.assign<entt::tag<"player"_hs>>(entity);
+        registry.assign<entt::tag<"game_scene"_hs>>(entity);
+        registry.assign<entt::tag<"dynamic"_hs>>(entity);
+
+        return entity;
+    }
+}
+
+// Column Logic System
+class column_logic final : public ecs::logic_update_system<column_logic> {
+public:
+    explicit column_logic(entt::registry &registry, entt::entity score_) noexcept : system(registry),
+                                                                                    score_entity(score_) {}
+    // Update, this will be called every tick
+    void update() noexcept final {
+        auto &registry = entity_registry_;
+
+        // Retrieve constants
+        const auto constants = registry.ctx<flappy_bird_constants>();
+
+        // Loop all columns
+        for (auto entity : registry.view<column>()) {
+            auto &col = registry.get<column>(entity);
+
```

Comment:
Here I would prefer separate each functionality of the loop in different functions, and Name it correctly, so the users Know exactly when reading the function, the purpose of the function.

Most of the time our functions should be < 25 lines of codes, if it's more, it's probably because we can separate

Reply:
For Example the function `move_pipe` is explicit and we know that we `move a pipe`

But in our `update()` function, we update different things, let's name a function for each things that we update.

Reply:
I think this part is really easy to read with the comments etc.